xoshiro256+x8.h

```diff
@@ -17,13 +19,32 @@ typedef struct prng_state {
   uint64_t state[4][XOSHIRO256_UNROLL];
 } prng_state;
 
+// Writes a 64-bit little endian integer to dst
+static inline void prng_write_le64(void *dst, uint64_t val) {
+  // Define to write in native endianness with memcpy
+  // Also, use memcpy on known little endian setups.
+# if defined(SHISHUA_NATIVE_ENDIAN) \
+   || defined(_WIN32) \
+   || (defined(__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__) \
+   || defined(__LITTLE_ENDIAN__)
+  memcpy(dst, &val, sizeof(uint64_t));
+#else
+  // Byteshift write.
+  uint8_t *d = (uint8_t *)dst;
+  for (size_t i = 0; i < 8; i++) {
+    d[i] = (uint8_t)(val & 0xff);
+    val >>= 8;
+  }
+#endif
+}
+
 // buf's size must be a multiple of 8 bytes.
-static inline void prng_gen(prng_state *s, uint64_t buf[], size_t size) {
+static inline void prng_gen(prng_state *s, uint8_t buf[], size_t size) {
   uint64_t t[XOSHIRO256_UNROLL];
-  size_t n = size / 8;
+  size_t n = size;
   uint64_t *b = (uint64_t *)buf;
-  for (size_t i = 0; i < n; i += XOSHIRO256_UNROLL) {
-    for (size_t j = 0; j < XOSHIRO256_UNROLL; j++) { buf[i + j] = s->state[0][j] + s->state[3][j]; }
+  for (size_t i = 0; i < n; i += XOSHIRO256_UNROLL * 8) {
+    for (size_t j = 0; j < XOSHIRO256_UNROLL; j++) { prng_write_le64(&buf[i + (8 * j)], s->state[0][j] + s->state[3][j]); }
```

Comment:
I’m not sure why, but the performance of this is severely degraded, much much slower than Xoshiro256+, when it should be much much faster.

Reply:
Hmm...perhaps it is compiling as a `vmovdqu` instead of `vmovdqa`.

Will have to check. What compiler and flags are you using?

Reply:
I just did `make test/benchmark-perf`.

I realize that a few important flags were lost. The initial command was:

    gcc -O9 -fdisable-tree-cunrolli -march=native -o xoshiro256plux8 prng.c

The new one:

    cc -O3 -g -DHEADER='"xoshiro256+x8.h"' prng.c -o xoshiro256plusx8